torch/csrc/jit/serialization/export_module.cpp

```diff
@@ -580,26 +489,43 @@ class ScriptModuleSerializer {
 
   void writeByteCode(const Module& module, const bool save_mobile_debug_info) {
     std::vector<c10::IValue> elements;
+    BackendDebugHandleManager debug_handle_manager;
     elements.emplace_back(
         static_cast<int64_t>(caffe2::serialize::kProducedBytecodeVersion));
-    c10::optional<std::vector<c10::IValue>> debug_info_elements;
-    if (save_mobile_debug_info) {
-      debug_info_elements = std::vector<c10::IValue>();
-      debug_info_elements->emplace_back(
-          static_cast<int64_t>(caffe2::serialize::kProducedBytecodeVersion));
-    }
+    std::vector<c10::IValue> debug_info_elements;
+    // Always save debug handles
+    debug_info_elements = std::vector<c10::IValue>();
+    debug_info_elements.emplace_back(
+        static_cast<int64_t>(caffe2::serialize::kProducedBytecodeVersion));
 
     moduleMethodsTuple(
-        module,
-        elements,
-        debug_info_elements,
-        save_mobile_debug_info,
-        source_range_tags_);
+        module, elements, debug_info_elements, debug_handle_manager);
     auto telements = Tup(std::move(elements));
     writeArchive("bytecode", telements);
+    auto debug_info_telements = Tup(std::move(debug_info_elements));
+    writeArchive("mobile_debug_handles", debug_info_telements);
```

Comment:
is the 'debug_info_telements' mobile specific? if so, it's unclear why it's not gated by 'save_mobile_debug_info' flag; if not, it's unclear why it's being written to 'mobile_debug_handles' archive.

Reply:
'debug_info_telements' is mobile specific, yes. It is not gated by save_mobile_debug_info because we chose to always save debug handles. `save_mobile_debug_info` is now used to select if we want to save callstack + source range info or not.

However saving only debug handle and not debug info is not quite coherent. We do want to save both together. And optionally strip off callstack stuff. But we will never strip off debug handles. So that is why I kept it this way. But to enable this we should really save both debug handles and debug info together and have utilities to strip debug info off.

I will make the change to save debug handles only if `save_mobile_debug_info`. This will make it coherent such that if we have debug handles we will have a way to symbolicate it.